addons/pkg-npm/jaxrs/src/main/java/org/commonjava/indy/pkg/npm/jaxrs/NPMContentAccessHandler.java

```diff
@@ -0,0 +1,424 @@
+import org.commonjava.indy.util.LocationUtils;
+import org.commonjava.maven.galley.TransferManager;
+import org.commonjava.maven.galley.event.EventMetadata;
+import org.commonjava.maven.galley.model.ConcreteResource;
+import org.commonjava.maven.galley.model.Transfer;
+import org.commonjava.maven.galley.model.TransferOperation;
+import org.commonjava.maven.galley.transport.htcli.model.HttpExchangeMetadata;
+import org.commonjava.maven.galley.util.PathUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.nio.file.Paths;
+import java.util.Date;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import static org.apache.commons.io.IOUtils.closeQuietly;
+import static org.commonjava.indy.bind.jaxrs.util.ResponseUtils.formatOkResponseWithEntity;
+import static org.commonjava.indy.bind.jaxrs.util.ResponseUtils.formatResponse;
+import static org.commonjava.indy.bind.jaxrs.util.ResponseUtils.setInfoHeaders;
+import static org.commonjava.indy.core.ctl.ContentController.LISTING_HTML_FILE;
+import static org.commonjava.indy.pkg.npm.util.GeneratedTransfers.getGeneratedTransfers;
+import static org.commonjava.indy.pkg.npm.util.GeneratedTransfers.setGeneratedTransfers;
+import static org.commonjava.maven.galley.spi.cache.CacheProvider.STORAGE_PATH;
+
+@ApplicationScoped
+@NPMContentHandler
+public class NPMContentAccessHandler
+        extends ContentAccessHandler
+{
+    protected final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    private TransferManager transfers;
+
+    @Inject
+    private ObjectMapper mapper;
+
+    @Override
+    public Response doCreate( String packageType, String type, String name, String path, HttpServletRequest request,
+                              EventMetadata eventMetadata, Supplier<URI> uriBuilder )
+    {
+        return doCreate( packageType, type, name, path, request, eventMetadata, uriBuilder, null );
+    }
+
+    @Override
+    public Response doCreate( String packageType, String type, String name, String path, HttpServletRequest request,
+                              EventMetadata eventMetadata, Supplier<URI> uriBuilder,
+                              Consumer<Response.ResponseBuilder> builderModifier )
+    {
+        final StoreType st = StoreType.get( type );
+        StoreKey sk = new StoreKey( packageType, st, name );
+
+        eventMetadata = eventMetadata.set( ContentManager.ENTRY_POINT_STORE, sk );
+
+        Response response = null;
+        final Transfer transfer;
+        try
+        {
+            InputStream stream = request.getInputStream();
+            transfer = contentController.store( sk, path, stream, eventMetadata );
```

Comment:
In our last discussion, I think we realized that storing this file up front and then re-reading it won't help our memory footprint. I think I'd prefer to see this code simply pass the stream into the generateNPMContentsFromTransfer() method, then have that method pass back the Transfer associated with the package.json you stored. 

Unless the version you're storing here is the merged version?? If that's the case, then ignore me. 😀

Reply:
@jdcasey, the reason I use transfer as param to pass is that we need to record the generated transfers from the target in the method: generateNPMContentsFromTransfer, and then after this generation, also need it to do the merging stuff, both of the generation and merging need its’ stream actually, I prefer to pass the transfer instead of copying any stream here and there for that.